### ISSUE TITLE
Issues in running preprocess

### ISSUE THREAD

**Issue by anupamme** [stanfordnlp/chirpycardinal#2]
Hello

I am trying to install the codebase and when I run preprocess.py like this

python3 preprocess.py ../../dump_dir/ ../../pgview_dir/ ../../wikidata_dir 24

This raises the following error:

File "preprocess.py", line 214, in stage0
    all_seeks_tup = [(args.dump_path, all_seeks[i], all_seeks[i + 1] - all_seeks[i]) for i in range(len(all_seeks) - 1)] + [(args.dump_path, all_seeks[-1], -1)]
IndexError: list index out of range

This is because the list all_seeks is empty ([]).

As per wiki-setup.md I set up all the variables. So I am not sure what is going wrong.

**Questions:**

Do I need to download the wiki dump manually and place it somewhere?

Also, is it possible to run it without spark as I am running this on a single-core machine? So that it is easier to debug.

And finally, thanks for the nice work and open-sourcing it.

**Comment by LithiumH** [comment]
Sorry for the late reply. You need to download 3 files and put them into the same directory.

1. Wikidata dump. Example: https://dumps.wikimedia.org/wikidatawiki/entities/20210215/ They can download the wikidata-20210215-all.json.bz2
2. Wikipedia dump. Example: https://dumps.wikimedia.org/enwiki/20210220/ they need to download enwiki-20210220-pages-articles-multistream.xml.bz2 and enwiki-20210220-pages-articles-multistream-index.txt.bz2
3. Pageview files, such as pagecounts-2020-08-views-ge-5-totals.bz2 from https://dumps.wikimedia.org/other/pagecounts-ez/merged/

You need to then supply the file names to the script. You can use the "--help" option to see it. If you run it on a single machine, Spark will also help with parallelizing it. If you run it on a single CPU it would be way too slow and may take days to preprocess. Let me know if you have other issues!

**Comment by anupamme** [comment]
Thanks for the clarification. With your instructions, I am able to run it.

But I am running it on a MacBook so it is 2 cores. Can you give an estimate of how long it takes to preprocess (build indexes) when run on an n-core machine (n=1 or higher)?

**Comment by anupamme** [comment]
Hello

Now I am running into another issue, while running (after a couple of hours of running) it throws following error. 

File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/util.py", line 107, in wrapper
    return f(*args, **kwargs)
  File "/Volumes/work/code/chirpycardinal/wiki-es-dump/preprocess.py", line 213, in <lambda>
    all_seeks = lines.map(lambda line: int(line.split(':', maxsplit=1)[0])).distinct().sortBy(lambda x:x).collect()
ValueError: invalid literal for int() with base 10: '      </contributor>'

File "/Volumes/work/code/chirpycardinal/wiki-es-dump/preprocess.py", line 213, in <lambda>
    all_seeks = lines.map(lambda line: int(line.split(':', maxsplit=1)[0])).distinct().sortBy(lambda x:x).collect()
ValueError: invalid literal for int() with base 10: '<mediawiki xmlns="http'

...

I am using:
python 3.7 
pyspark 3.0.2
macOS 10.13.6 (High Sierra).

It seems there is some encoding issue. Would it be possible to give some pointers what can be the possible reasons for this error.

**Comment by AshwinParanjape** [comment]
Looking at the error, it seems that it is trying to convert `' '` and `'<mediawiki xmlns="http'` into an int. 

Looking at the code around line 213
```python3

def stage0(args):
    """
    This mega stage parses everything
    """
    print(f'>  Processing Wiki dump at {args.dump_path}')
    index_path = args.dump_path.replace('.xml', '-index.txt')
    lines = sc.textFile(index_path).repartition(256)
    all_seeks = lines.map(lambda line: int(line.split(':', maxsplit=1)[0])).distinct().sortBy(lambda x:x).collect()
    all_seeks_tup = [(args.dump_path, all_seeks[i], all_seeks[i + 1] - all_seeks[i]) for i in range(len(all_seeks) - 1)] + [(args.dump_path, all_seeks[-1], -1)]
    all_seeks_tup = sc.parallelize(all_seeks_tup).repartition(2048)
```

It seems to be taking the dump_path, which ends in '.xml', replaces it with '.index.txt' and then for every line in it, splits at colon and takes the first part as an integer. Looking at the 'enwiki-20210220-pages-articles-multistream-index.txt.index.txt' it has a fixed format which always begins with an integer. Greping for 'mediawiki' leads lines that always begin with int as well. 

So I think what might have happened here is that you might be reading the xml dump instead of the index.txt file. Can you confirm that you haven't made any such changes to the code? Can you also post the command you are running (along with the arguments?)

**Comment by anupamme** [comment]
I have not made any changes to the code, you are right that the error happens on that line.

The command I use to run is:

spark-submit --conf spark.local.dir=/Volumes/work/new_data/chirpy/spark/ --driver-memory 50G preprocess.py /Volumes/work/new_data/chirpy/dump/ /Volumes/work/new_data/chirpy/pgview/ /Volumes/work/new_data/chirpy/wiki/ 24

When I run 

cat enwiki-20210220-pages-articles-multistream-index.txt | grep mediawiki 

The output I get is:

1506363006:686356:MediaWiki:Nstab-mediawiki
5969941600:11337457:File:Possiblemediawikibug.JPG
6964063153:15044680:Wikipedia:WikiProject Spam/LinkSearch/asianmediawiki.com
6988396244:15140001:Wikipedia:WikiProject Spam/LinkReports/asianmediawiki.com
7711043201:18348165:Template:User mediawiki
8518423574:21645446:File:Poweredby mediawiki 88x311.png
8526132421:21678988:File:Poweredby mediawiki 88x311 6.png
8561316393:21821424:Semantic mediawiki
9907446002:27078364:List of mediawiki sites
13300736717:40880390:Wikipedia:Articles for deletion/Wikipmediawiki
17651182793:60025969:M.mediawiki.org

All of them begin with an integer. The dump directory contains two files:

enwiki-20210220-pages-articles-multistream-index.txt
enwiki-20210220-pages-articles-multistream.xml

Is it possible that this is due to a newer version of wiki dump (20210220)?

**Comment by anupamme** [comment]
The code is indeed reading the xml file, here is the trace:

`**21/03/02 23:03:48 INFO HadoopRDD: Input split: file:/Volumes/work/new_data/chirpy/dump/enwiki-20210220-pages-articles-multistream.xml:82074140672+10512034**
21/03/02 23:03:51 INFO PythonRunner: Times: total = 3113, boot = -344, init = 346, finish = 3111
21/03/02 23:03:52 INFO Executor: Finished task 2476.0 in stage 0.0 (TID 2476). 2110 bytes result sent to driver
21/03/02 23:03:52 INFO TaskSetManager: Finished task 2476.0 in stage 0.0 (TID 2476) in 3242 ms on 192.168.0.11 (executor driver) (2474/2477)
21/03/02 23:03:53 INFO PythonRunner: Times: total = 7868, boot = -332, init = 335, finish = 7865
21/03/02 23:03:53 INFO Executor: Finished task 2473.0 in stage 0.0 (TID 2473). 2110 bytes result sent to driver
21/03/02 23:03:53 INFO TaskSetManager: Finished task 2473.0 in stage 0.0 (TID 2473) in 8198 ms on 192.168.0.11 (executor driver) (2475/2477)
21/03/02 23:03:54 INFO PythonRunner: Times: total = 8395, boot = -328, init = 332, finish = 8391
21/03/02 23:03:54 INFO Executor: Finished task 2474.0 in stage 0.0 (TID 2474). 2110 bytes result sent to driver
21/03/02 23:03:54 INFO TaskSetManager: Finished task 2474.0 in stage 0.0 (TID 2474) in 8723 ms on 192.168.0.11 (executor driver) (2476/2477)
21/03/02 23:03:55 INFO PythonRunner: Times: total = 7532, boot = -500, init = 502, finish = 7530
21/03/02 23:03:56 INFO Executor: Finished task 2475.0 in stage 0.0 (TID 2475). 2110 bytes result sent to driver
21/03/02 23:03:56 INFO TaskSetManager: Finished task 2475.0 in stage 0.0 (TID 2475) in 7826 ms on 192.168.0.11 (executor driver) (2477/2477)
21/03/02 23:03:56 INFO TaskSchedulerImpl: Removed TaskSet 0.0, whose tasks have all completed, from pool 
21/03/02 23:03:56 INFO DAGScheduler: ShuffleMapStage 0 (coalesce at NativeMethodAccessorImpl.java:0) finished in 5306.008 s
21/03/02 23:03:56 INFO DAGScheduler: looking for newly runnable stages
21/03/02 23:03:56 INFO DAGScheduler: running: Set()
21/03/02 23:03:56 INFO DAGScheduler: waiting: Set(ShuffleMapStage 1, ResultStage 2)
21/03/02 23:03:56 INFO DAGScheduler: failed: Set()
21/03/02 23:03:56 INFO DAGScheduler: Submitting ShuffleMapStage 1 (PairwiseRDD[8] at distinct at /Volumes/work/code/chirpycardinal/wiki-es-dump/preprocess.py:213), which has no missing parents
21/03/02 23:03:56 INFO MemoryStore: Block broadcast_2 stored as values in memory (estimated size 11.1 KiB, free 29.8 GiB)
21/03/02 23:03:56 INFO MemoryStore: Block broadcast_2_piece0 stored as bytes in memory (estimated size 6.7 KiB, free 29.8 GiB)
21/03/02 23:03:56 INFO BlockManagerInfo: Added broadcast_2_piece0 in memory on 192.168.0.11:53355 (size: 6.7 KiB, free: 29.8 GiB)
21/03/02 23:03:56 INFO SparkContext: Created broadcast 2 from broadcast at DAGScheduler.scala:1223
21/03/02 23:03:56 INFO DAGScheduler: Submitting 256 missing tasks from ShuffleMapStage 1 (PairwiseRDD[8] at distinct at /Volumes/work/code/chirpycardinal/wiki-es-dump/preprocess.py:213) (first 15 tasks are for partitions Vector(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14))
21/03/02 23:03:56 INFO TaskSchedulerImpl: Adding task set 1.0 with 256 tasks
21/03/02 23:03:56 INFO TaskSetManager: Starting task 0.0 in stage 1.0 (TID 2477, 192.168.0.11, executor driver, partition 0, PROCESS_LOCAL, 7408 bytes)
21/03/02 23:03:56 INFO TaskSetManager: Starting task 1.0 in stage 1.0 (TID 2478, 192.168.0.11, executor driver, partition 1, PROCESS_LOCAL, 7408 bytes)
21/03/02 23:03:56 INFO TaskSetManager: Starting task 2.0 in stage 1.0 (TID 2479, 192.168.0.11, executor driver, partition 2, PROCESS_LOCAL, 7408 bytes)
21/03/02 23:03:56 INFO TaskSetManager: Starting task 3.0 in stage 1.0 (TID 2480, 192.168.0.11, executor driver, partition 3, PROCESS_LOCAL, 7408 bytes)
21/03/02 23:03:56 INFO Executor: Running task 3.0 in stage 1.0 (TID 2480)
21/03/02 23:03:56 INFO Executor: Running task 1.0 in stage 1.0 (TID 2478)
21/03/02 23:03:56 INFO Executor: Running task 2.0 in stage 1.0 (TID 2479)
21/03/02 23:03:56 INFO Executor: Running task 0.0 in stage 1.0 (TID 2477)
21/03/02 23:03:56 INFO ShuffleBlockFetcherIterator: Getting 2477 (194.2 MiB) non-empty blocks including 2477 (194.2 MiB) local and 0 (0.0 B) host-local and 0 (0.0 B) remote blocks
21/03/02 23:03:56 INFO ShuffleBlockFetcherIterator: Getting 2477 (194.5 MiB) non-empty blocks including 2477 (194.5 MiB) local and 0 (0.0 B) host-local and 0 (0.0 B) remote blocks
21/03/02 23:03:56 INFO ShuffleBlockFetcherIterator: Getting 2477 (194.7 MiB) non-empty blocks including 2477 (194.7 MiB) local and 0 (0.0 B) host-local and 0 (0.0 B) remote blocks
21/03/02 23:03:56 INFO ShuffleBlockFetcherIterator: Getting 2477 (194.9 MiB) non-empty blocks including 2477 (194.9 MiB) local and 0 (0.0 B) host-local and 0 (0.0 B) remote blocks
21/03/02 23:03:56 INFO ShuffleBlockFetcherIterator: Started 0 remote fetches in 72 ms
21/03/02 23:03:56 INFO ShuffleBlockFetcherIterator: Started 0 remote fetches in 72 ms
21/03/02 23:03:56 INFO ShuffleBlockFetcherIterator: Started 0 remote fetches in 73 ms
21/03/02 23:03:56 INFO ShuffleBlockFetcherIterator: Started 0 remote fetches in 73 ms
/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/shuffle.py:60: UserWarning: Please install psutil to have better support with spilling
/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/shuffle.py:60: UserWarning: Please install psutil to have better support with spilling
/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/shuffle.py:60: UserWarning: Please install psutil to have better support with spilling
/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/shuffle.py:60: UserWarning: Please install psutil to have better support with spilling
21/03/02 23:03:58 ERROR Executor: Exception in task 3.0 in stage 1.0 (TID 2480)
org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 605, in main
    process()
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 595, in process
    out_iter = func(split_index, iterator)
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 2596, in pipeline_func
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 2596, in pipeline_func
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 425, in func
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 1946, in combineLocally
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/shuffle.py", line 238, in mergeValues
    for k, v in iterator:
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/util.py", line 107, in wrapper
    return f(*args, **kwargs)
  File "/Volumes/work/code/chirpycardinal/wiki-es-dump/preprocess.py", line 213, in <lambda>
    all_seeks = lines.map(lambda line: int(line.split(':', maxsplit=1)[0])).distinct().sortBy(lambda x:x).collect()
ValueError: invalid literal for int() with base 10: '      <namespace key="0" case="first-letter" />'
`

**Comment by AshwinParanjape** [comment]
Can you change the wikipedia and wikidata paths from folders to files? I think the instructions are erroneous, in that they ask you to provide a folder path and not the file path. 
i.e. 
/Volumes/work/new_data/chirpy/dump/ ->/Volumes/work/new_data/chirpy/dump/enwiki-20210220-pages-articles-multistream.xml (or .xml.bz2 if it is compressed)
/Volumes/work/new_data/chirpy/wiki/ ->/Volumes/work/new_data/chirpy/wiki/wikidata-20210215-all.json.bz2 

(If the filenames don't match exactly, just make sure they point to the files you have on your system)

**Comment by anupamme** [comment]
I tried with using the filenames instead of folders so the new command is 

spark-submit --conf spark.local.dir=/Volumes/work/new_data/chirpy/spark/ --driver-memory 50G preprocess.py /Volumes/work/new_data/chirpy/dump/enwiki-20210220-pages-articles-multistream.xml /Volumes/work/new_data/chirpy/pgview/pagecounts-2020-08-views-ge-5-totals /Volumes/work/new_data/chirpy/wiki/wikidata-20210215-all.json.bz2 24

However, now I am getting a different error:

21/03/03 00:37:44 ERROR Utils: Aborting task
org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 605, in main
    process()
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 597, in process
    serializer.dump_stream(out_iter, outfile)
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/serializers.py", line 141, in dump_stream
    for obj in iterator:
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 1644, in func
  File "/Volumes/work/code/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/util.py", line 107, in wrapper
    return f(*args, **kwargs)
  File "/Volumes/work/code/chirpycardinal/wiki-es-dump/preprocess.py", line 55, in process_seek
    xml = bz2.decompress(doc).decode('utf-8').replace('</mediawiki>', '')
  File "/usr/local/Cellar/python/3.7.6_1/Frameworks/Python.framework/Versions/3.7/lib/python3.7/bz2.py", line 346, in decompress
    res = decomp.decompress(data)
OSError: Invalid data stream

It seems that there is something incorrect the way I am running it.

**Comment by AshwinParanjape** [comment]
I think it is expecting a bz2 compressed file there. This was my mistake.

Instead of 
/Volumes/work/new_data/chirpy/dump/enwiki-20210220-pages-articles-multistream.xml

can you supply the bz2 compressed file? 
/Volumes/work/new_data/chirpy/dump/enwiki-20210220-pages-articles-multistream.xml.bz2

The string replace on the argument path should still work (by converting xml.bz2 to -index.txt.bz2)

**Comment by anupamme** [comment]
By changing it to bz2 the command now works.

So the preprocess.py has been running for the last few hours. Is there an estimate of how long it will take on a 2 core laptop?

Since there are no further questions I am closing the issues. 

The proactive response was much appreciated.

**Comment by anupamme** [comment]
I am now running into issues in using the upload.py. Perhaps I should have opened a new issue but then it is related to this one so I am just reopening this one.

I was able to run the preprocess.py and it created output in the dump_path:

1. enwiki-20210220-pages-articles-multistream-articles.json.bz2
2. enwiki-20210220-pages-articles-multistream-wikidata-integrated.json.bz2

However, in the upload.py it says in line 31 and 32 say:

Fully Qualified path of the processed *-sections.json.bz2 file
Fully Qualified path of the *-integrated.json.bz2 file

Q1: I did not find any *-sections.json.bz2 file in the dump_folder, however, I did find *-integrated.json.bz2 file and *-articles.json.bz2. So do I need to run the preprocess again? 

I looked at preprocess.py and I could not find which stage prepares the *-sections.json.bz2 file, only line I found is line 272 which mentions -sections.json, but that is also commented out.

So can you please guide what am I missing here?

Q2: I tried running upload.py with only articles_path by commenting out line 44 (which uses sections_path) and I am getting following error on line 47 (sc.textFile(args.articles_path)):

objc[12657]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called.
objc[12657]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
objc[12659]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called.
objc[12659]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
objc[12658]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called.
objc[12658]: +[__NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
21/03/21 12:12:02 ERROR PythonRunner: Python worker exited unexpectedly (crashed)
java.net.SocketException: Connection reset
	at java.base/java.net.SocketInputStream.read(SocketInputStream.java:210)
	at java.base/java.net.SocketInputStream.read(SocketInputStream.java:141)
	at java.base/java.io.BufferedInputStream.fill(BufferedInputStream.java:252)
	at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:271)

I am not sure why multiple forks are getting called. So this is my second question.

I tried using both spark-submit and python3 upload.py with arguments but I get the same error.

If you need any more information please let me know.

**Comment by AshwinParanjape** [comment]
@LithiumH wrote the code for es preprocessing, so he would have to weigh in on this one. 

Q1: Looking at https://github.com/stanfordnlp/chirpycardinal/blob/955a14380b3ff43b1c567af6a1cd9fdf81fb7ece/wiki-es-dump/preprocess.py#L272

I agree with you. It seems that the code to generate sections.json.bz2 is commented out. I'm pretty sure it is a mistake and it shouldn't be commented out. I haven't looked this up, but perhaps the bz2 file gets created because the bz2 codec is being used while calling `saveAsTextFile`, so even if the path says sections.json, you would get sections.json.bz2 out of it. 

For @anupamme if you haven't deleted the intermediate files, in particular a file ending with parsed.json(.bz2), rerunning stage0 and entire stage 1 might be unnecessary and time consuming. I would recommend commenting out lines 230 and 233-270 and modifying line 376 to run only stage 1. 

Q2: This is very odd. Googling for the error took me to a lot places where people had similar issue with Ruby app servers. One comprehensive explanation is here https://blog.phusion.nl/2017/10/13/why-ruby-app-servers-break-on-macos-high-sierra-and-what-can-be-done-about-it/. It seems to be a problem because MacOS throws an error instead of failing silently. 

I believe we've never run these processing scripts from a mac device which is why we didn't encounter this error. The suggested fix is to `export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES`. Let me know if this works for you.

**Comment by anupamme** [comment]
On the issue or **creating -sections.json.bz2 file**:
I ran the preprocess after commenting out the lines as you suggested and now I have -sections.json.bz2 file as an output. So this part is sorted for me.

On the issue of **running upload.py**:
Thanks for pointing me to the mac related issues. I tried running upload.py on a newer OS X version: 10.15.7 (Catalina). And I did not get the same error, however, I got a different error (elasticsearch.helpers.errors.BulkIndexError). On that, I wanted to check which version of elasticsearch are you guys using?

My versions: elasticsearch (from brew) - 7.10.2 and elasticsearch (from pip) 7.5.1.

I also checked by querying elasticsearch for some common words: hello, coffee, tea, however I get empty results. The query I make is:
request: es.search(index="enwiki-20201201-articles", body={'query': {'bool': {'should': [{'terms': {'linkable_span': ['tea']}}]}}, 'sort': {'pageview': 'desc'}}, size=1000, request_timeout=1000000)
response: {'took': 22, 'timed_out': False, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}, 'hits': {'total': {'value': 0, 'relation': 'eq'}, 'max_score': None, 'hits': []}}

I also ran a query to get all the indices:
request: es.indices.get_alias("*")
response: {'enwiki-20201201-sections': {'aliases': {}}, 'enwiki-20201201-articles': {'aliases': {}}}

So if you could tell me the version numbers of elastic search you are using? Also, which OS are you using elastic search on?

Here is more trace of the error:

21/03/23 13:58:21 INFO SparkUI: Stopped Spark web UI at http://192.168.0.20:4040
21/03/23 13:58:21 INFO TaskSchedulerImpl: Removed TaskSet 0.0, whose tasks have all completed, from pool 
21/03/23 13:58:21 WARN TaskSetManager: Lost task 2.0 in stage 0.0 (TID 2) (192.168.0.20 executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/Users/laptop/code/ChatBot/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 604, in main
    process()
  File "/Users/laptop/code/ChatBot/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 594, in process
    out_iter = func(split_index, iterator)
  File "/Users/laptop/code/ChatBot/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 2916, in pipeline_func
  File "/Users/laptop/code/ChatBot/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 2916, in pipeline_func
  File "/Users/laptop/code/ChatBot/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 2916, in pipeline_func
  [Previous line repeated 1 more time]
  File "/Users/laptop/code/ChatBot/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 418, in func
  File "/Users/laptop/code/ChatBot/chirpycardinal/chirpy/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 932, in func
  File "/Users/laptop/code/ChatBot/chirpycardinal/wiki-es-dump/upload.py", line 24, in upload_partition
    bulk(es, actions)
  File "/Users/laptop/code/ChatBot/chirpycardinal/chirpy/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 304, in bulk
    for ok, item in streaming_bulk(client, actions, *args, **kwargs):
  File "/Users/laptop/code/ChatBot/chirpycardinal/chirpy/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 234, in streaming_bulk
    **kwargs
  File "/Users/laptop/code/ChatBot/chirpycardinal/chirpy/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 162, in _process_bulk_chunk
    raise BulkIndexError("%i document(s) failed to index." % len(errors), errors)
elasticsearch.helpers.errors.BulkIndexError:

**Comment by AshwinParanjape** [comment]
elasticsearch version 7.8
python elasticsearch version 7.0.5 (pip)

Looking at the errors, I don't think the upload process uploaded any documents. We are using `bulk` to upload the documents. 

```raise BulkIndexError("%i document(s) failed to index." % len(errors), errors)```
Is there any missing output, this line tells me that it should have printed errors for the failure of Bulk Indexing. 

Few things you can try
* Count the number of documents currently uploaded. There is an API call for that. 
* Try uploading a single dummy document (you would have to look at elasticsearch python API for that).

**Comment by anupamme** [comment]
Hello

Sorry for the late response. The issue was that I did not have enough space on the disk therefore the BulkIndexError was coming.

**Question:** To fix this, I have to upgrade my hard drive. I have **an external hard drive**, is it possible to use that to run **upload.py**. And point spark to use a custom link on the file system.

I am also trying to upgrade my internal SSD but given it is an older Mac it is a bit non-trivial.

**Comment by AshwinParanjape** [comment]
> I have an external hard drive, is it possible to use that to run upload.py. And point spark to use a custom link on the file system.

Yeah, that should be possible. The easiest would probably be to move `/Volumes/work/new_data/chirpy/dump/` and create a symlink to it. I'm assuming this folder is occupying a lot of space.

**Comment by anupamme** [comment]
I am still not able to resolve the BulkIndexError (trace is attached below). 

I tried increasing the hard disk on my local to 1 TB (so hard disk space is not a problem).
The RAM is 16 GB and that is also not a problem.
I also took 10% of the data (files starting with part-00*) in both directories: -sections.json.bz2 and -integrated.json.bz2 (to reduce the memory and space requirements).

Once the error is done I ran some queries on elasticsearch to see how much data has been uploaded and here is the output:

1. query: es.indices.get_alias("*")
response: {'enwiki-20201201-articles': {'aliases': {}}, 'test-index': {'aliases': {}}, 'enwiki-20201201-sections': {'aliases': {}}}

2. query: res = es.search(index="enwiki-20201201-articles", body={"query": {"match_all": {}}})
res['hits']['total']['value'] is 10000
len(res['hits']['hits']) is 10

3. query: res = es.search(index="enwiki-20201201-sections", body={"query": {"match_all": {}}})
res['hits']['total']['value'] is 10000
len(res['hits']['hits']) is 10

So I have run out of options to try so would appreciate it if you could suggest how to fix this.

Command to run: spark-submit --conf spark.local.dir=/Users/mediratta/data/chirpy/spark --driver-memory 200G upload.py -ap /Users/mediratta/data/chirpy/enwiki-20210220-pages-articles-multistream-wikidata-integrated.json.bz2/sub -sp /Users/mediratta/data/chirpy/enwiki-20210220-pages-articles-multistream-sections.json.bz2/sub -U '' -P '' -d localhost -p 9200

Here is the complete stack trace:

File "/usr/local/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 604, in main
    process()
  File "/usr/local/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 594, in process
    out_iter = func(split_index, iterator)
  File "/usr/local/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 2916, in pipeline_func
  File "/usr/local/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 2916, in pipeline_func
  File "/usr/local/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 2916, in pipeline_func
  [Previous line repeated 1 more time]
  File "/usr/local/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 418, in func
  File "/usr/local/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 932, in func
  File "/Users/mediratta/code/chirpycardinal/wiki-es-dump/upload.py", line 25, in upload_partition
    bulk(es, actions)
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 396, in bulk
    for ok, item in streaming_bulk(client, actions, *args, **kwargs):
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 326, in streaming_bulk
    **kwargs
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 246, in _process_bulk_chunk
    for item in gen:
  File "/usr/local/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 185, in _process_bulk_chunk_success
    raise BulkIndexError("%i document(s) failed to index." % len(errors), errors)
elasticsearch.helpers.errors.BulkIndexError: ('375 document(s) failed to index.', [{'index': {'_index': 'enwiki-20201201-articles'

**Comment by AshwinParanjape** [comment]
I'm not 100% sure that the following is actually indicative of the number of documents. 
```
res['hits']['total']['value'] is 10000
len(res['hits']['hits']) is 10
```

Can you try - https://elasticsearch-py.readthedocs.io/en/v7.12.0/api.html#elasticsearch.Elasticsearch.count
My hunch is that a few documents are not in the right format for some reason and got rejected, but if res['hits']['total']['value'] is 10000 then a lot of documents should have gotten uploaded and what got rejected might be small enough for you to ignore. 

FYI - ES by default only returns 10 documents per query, so no need to worry about that. And there's a cap of 10000 documents that it scores exactly. You can check the documentation https://www.elastic.co/guide/en/elasticsearch/reference/7.12/search-search.html
This might explain the two numbers that you are seeing above.

**Comment by anupamme** [comment]
Okay, I think there is no good way to address the error so I was wondering if this is possible for you guys:

Can you make the output of process.py available for download (in some cloud storage like S3 bucket)? 

Because I would like to separate out the cause of the bug: was it some incorrect document format (which can be validated if I can use a set of documents that work in your case) or some issue in the upload.py (which is causing the BulkIndexError).

**Comment by AshwinParanjape** [comment]
But could you count the number of documents that got uploaded? using the `count` function call? 

Tagging @LithiumH to help with this. He might have some insights or access to the dump.

**Comment by anupamme** [comment]
Here is the count output

es.count(index="enwiki-20201201-sections")
{'count': 154258473, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}}

es.count(index="enwiki-20201201-articles")
{'count': 236229, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}}

Does this sound like what you'd expect?

**Comment by AshwinParanjape** [comment]
That's a pretty large number, which is what we expect! You can compare with the number of articles and sections in the json file that you tried to upload and see what is the difference. 

Just to double check - When you used `part-00` was it 10% or 1%? The 00 prefix leads me to believe that it could be 1%.

**Comment by anupamme** [comment]
I tried both 10% and 1%: and I was getting BulkIndexError for both the cases.

This tells me that it is not really the amount of data which is causing the issue.

If the number of documents is about what you expect and we don't know why this error is coming then perhaps we can close this issue as I don't know what we can do next.

**Comment by AshwinParanjape** [comment]
Yeah, I suspect the issue really boils down too some processing error or some sort of corruption with some of the documents. And so I think apart from those docs, everything else is probably getting indexed. 

The wikipedia index is used for linking entities and in some cases talking about them. If while running the bot you see a lot of errors complaining about missing indices in elasticsearch, then we can drill further.

**Comment by anupamme** [comment]
I am getting lot of elastic search missing errors when I run the bot so I am re-opening it. Please bear with me.

I following steps:
1. I removed the earlier indices to start fresh (after this step both indices: enwiki-20201201-sections and enwiki-20201201-articles were absent).
2. I ran upload.py, this time only 1 file each for sections_path (part-00100.bz2) and articles_path (part-00100.bz2).
 
But I get the same BulkIndexError (stack is attached). It says that 382 documents were failed to index.

However, when I run
>es.count(index="enwiki-20201201-sections")
{'count': 18570, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}}
> es.count(index="enwiki-20201201-articles")
{'count': 221, '_shards': {'total': 1, 'successful': 1, 'skipped': 0, 'failed': 0}}

Q1: This means that 221 documents were indexed. Since 382 failed to index, does 1 bz2 file contains around 600 documents, is this number correct? I am just trying to understand the file structure.

Q2: Can you guys provide just 1 bz2 file for each section and article, which has worked in your case. I want to investigate whether the bz2 files I have are corrupted or there is some issue in the code or local setup.



21/05/18 17:49:09 INFO HadoopRDD: Input split: file:/Users/mediratta/data/chirpy/enwiki-20210220-pages-articles-multistream-wikidata-integrated.json.bz2/sub/sub/sub/part-00100.bz2:0+3212098
21/05/18 17:49:09 INFO HadoopRDD: Input split: file:/Users/mediratta/data/chirpy/enwiki-20210220-pages-articles-multistream-wikidata-integrated.json.bz2/sub/sub/sub/part-00100.bz2:3212098+3212099
21/05/18 17:49:10 INFO BlockManagerInfo: Removed broadcast_1_piece0 on 192.168.0.20:59017 in memory (size: 5.6 KiB, free: 119.8 GiB)
21/05/18 17:49:10 ERROR Executor: Exception in task 0.0 in stage 1.0 (TID 2)
org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "/Users/mediratta/code/chirpycardinal/ch/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 604, in main
    process()
  File "/Users/mediratta/code/chirpycardinal/ch/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/worker.py", line 594, in process
    out_iter = func(split_index, iterator)
  File "/Users/mediratta/code/chirpycardinal/ch/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 2916, in pipeline_func
  File "/Users/mediratta/code/chirpycardinal/ch/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 2916, in pipeline_func
  File "/Users/mediratta/code/chirpycardinal/ch/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 2916, in pipeline_func
  [Previous line repeated 1 more time]
  File "/Users/mediratta/code/chirpycardinal/ch/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 418, in func
  File "/Users/mediratta/code/chirpycardinal/ch/lib/python3.7/site-packages/pyspark/python/lib/pyspark.zip/pyspark/rdd.py", line 932, in func
  File "/Users/mediratta/code/chirpycardinal/wiki-es-dump/upload.py", line 24, in upload_partition
    bulk(es, actions)
  File "/Users/mediratta/code/chirpycardinal/ch/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 396, in bulk
    for ok, item in streaming_bulk(client, actions, *args, **kwargs):
  File "/Users/mediratta/code/chirpycardinal/ch/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 326, in streaming_bulk
    **kwargs
  File "/Users/mediratta/code/chirpycardinal/ch/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 246, in _process_bulk_chunk
    for item in gen:
  File "/Users/mediratta/code/chirpycardinal/ch/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 185, in _process_bulk_chunk_success
    raise BulkIndexError("%i document(s) failed to index." % len(errors), errors)
elasticsearch.helpers.errors.BulkIndexError: ('382 document(s) failed to index.', [{'index': {'_index': 'enwiki-20201201-articles', '_type': '_doc', '_id':

**Comment by anupamme** [comment]
I investigated it a bit further:

I unzipped the file part-00100.bz2 (in both sections and articles) and its size (number of lines) was:

articles: 6086
sections: 18570

In order to get to there the issue is in the bz2 files: I started running upload.py on subsets of both files (first 1/2, first 1/4, ..), until I was left with the following number of lines in each file (in every try I was getting the same BulkIndexError)

articles: 6
sections: 18 

In this case, I again got BullIndexError but it specifies the exact document (1 document is failed to index) (trace is attached).

I am a bit clueless why this error is coming. I wanted to check that by looking at the document (in articles index) which is failed to index, does it give any idea of what is probably happening?

P.S: I am not able to re-open this issue but if it helps I can open a new issue :).

raise BulkIndexError("%i document(s) failed to index." % len(errors), errors)
elasticsearch.helpers.errors.BulkIndexError: ('1 document(s) failed to index.', [{'index': {'_index': 'enwiki-20201201-articles', '_type': '_doc', '_id': 'XXDaiHkBjw_GT6rSYQAX', 'status': 400, 'error': {'type': 'illegal_argument_exception', 'reason': 'mapper [linkable_span_info] cannot be changed from type [long] to [text]'}, 'data': {'doc_title': 'Óscar Aguilera', 'doc_id': '33272394', 'categories': ['1935 births', 'Paraguayan footballers', 'Paraguay international footballers', 'Association football forwards', 'Club Olimpia footballers', '1958 FIFA World Cup players', 'Living people'], 'redirects': ['Oscar Aguilera', 'Óscar Aguilera'], 'pageview': 24, 'overview_section': 'Óscar Antonio Aguilera Valdés :es:Óscar Aguilera (born 11 March 1935) is a Paraguayan football forward who played for Paraguay in the 1958 FIFA World Cup. He also played for Club Olimpia.', 'full_text': 'Óscar Antonio Aguilera Valdés :es:Óscar Aguilera (born 11 March 1935) is a Paraguayan football forward who played for Paraguay in the 1958 FIFA World Cup. He also played for Club Olimpia.\n\nReferences\n\nExternal links', 'linkable_span': ['aguilera'], 'linkable_span_info': [('aguilera', 5)], 'wikidata_id': 'Q2058563', 'wikidata_categories_all': ['legal person', 'subject of a right', 'human', 'organism', 'player', 'omnivore', 'anatomical entity', 'sports figure', 'organisms by adaptation', 'heterotroph', 'consumer', 'set', 'agent', 'entity', 'independent continuant', 'collection entity', 'association football player', 'athlete', 'natural person', 'item with given name property', 'individual entity', 'continuant', 'group', 'subject', 'person', 'competitive player', 'individual'], 'wikidata_categories_info': '{"legal person": 2, "subject of a right": 3, "human": 0, "organism": 3, "player": 2, "omnivore": 1, "anatomical entity": 4, "sports figure": 2, "organisms by adaptation": 2, "heterotroph": 3, "consumer": 2, "set": 4, "agent": 2, "entity": 3, "independent continuant": 5, "collection entity": 5, "association football player": 0, "athlete": 1, "natural person": 1, "item with given name property": 2, "individual entity": 3, "continuant": 6, "group": 3, "subject": 2, "person": 1, "competitive player": 1, "individual": 2}'}}}])

**Comment by anupamme** [comment]
More digging, there is an error key in the stack trace, whose value is:

{'type': 'illegal_argument_exception', 'reason': 'mapper [linkable_span_info] cannot be changed from type [long] to [text]'}

**'reason': 'mapper [linkable_span_info] cannot be changed from type [long] to [text]'**

Does it tell which field is incorrect?

**Comment by AshwinParanjape** [comment]
That is some excellent sleuthing! Kudos 👍 

I see that the type is defined as text here:
https://github.com/stanfordnlp/chirpycardinal/blob/635957802720467cd4447caa9481ad6152f10bd3/wiki-es-dump/define_es.py#L20

I suspect that in your case when you started fresh the schema wasn't defined and it was inferred (perhaps based on the first doc) that linkable span info is an integer (long) and not text. 

Can you do:
(As you did before) remove the earlier indices to start fresh. 
Then run define_es and make sure the types are defined.
Then try uploading the 6 documents. 

PS - It might be confusing to see a tuple stored as text, but es doesn't really support nested data structures, so text is our best bet.

**Comment by anupamme** [comment]
Yes, following these steps helped fix the issue so I guess what was missing earlier was running defined_es.py (it is there in wiki-setup.md, somehow I missed it).

After fixing this issue I tried running upload on a larger dataset and ran into a new issue which is:

'error': {'type': 'illegal_argument_exception', 'reason': 'Document contains at least one immense term in field="linkable_span" (whose UTF8 encoding is longer than the max length 32766), all of which were skipped.  Please correct the analyzer to not produce such terms.  The prefix of the first immense term is: \'[124, 32, 116, 111, 112, 32, 115, 99, 111, 114, 101, 52, 32, 61, 32, 50, 54, 48, 32, 124, 32, 100, 101, 108, 105, 118, 101, 114, 105, 101]...\', original message: bytes can be at most 32766 in length; got 35101', 'caused_by': {'type': 'max_bytes_length_exceeded_exception', 'reason': 'bytes can be at most 32766 in length; got 35101'}}

This means that the field **linkable_span** has exceeded its max size. Here is the document for which this error is coming and the **linkable_span** is indeed quite big. Any ideas on how to handle this error?

The file containing this document with a large **linkable_span** is attached.

[linkable_span_large.txt](https://github.com/stanfordnlp/chirpycardinal/files/6521245/linkable_span_large.txt)

**Comment by AshwinParanjape** [comment]
Huh, so the linkable spans are anchors in the wikipedia article. It seems like in this article about cricket a lot numbers were linked. In that case, I don't think it is much of a problem if the linkable span for this article is just dropped from the index. 

Can you confirm other documents were successfully uploaded? If so, there's no need to fix this doc, it won't really affect the performance of the bot. 

If you really wanted to fix this error, you would have to make a choice about which spans to keep (say from the top of the page or something) and run the preprocessor to chop off at a certain length. But that seems like a lot of work for little gain.

**Comment by anupamme** [comment]
So it uploads the upload routine uploads the documents till there is not error/exception, once there is an exception it throws the exception and stops the execution, since the exceptions are not caught (upload.py, line 24).

So the documents are uploaded till it encounters this document. So in order to upload all the documents, I need to identify which file contains this document and remove it from there. Or is there a better way (I will also try catching the error to see if that works).

I do not want to fix this error I just want to remove/ignore the problematic document.

**Comment by AshwinParanjape** [comment]
Catching the error is better, otherwise you'll have to do document removal manually. It seems that we didn't encounter this error before. 

You would have to add a try-except here:
https://github.com/stanfordnlp/chirpycardinal/blob/635957802720467cd4447caa9481ad6152f10bd3/wiki-es-dump/upload.py#L24

If that works for you, I would appreciate it if you could submit a PR! :)

**Comment by anupamme** [comment]
> Sorry for the late reply. You need to download 3 files and put them into the same directory.
> 
> 1. Wikidata dump. Example: https://dumps.wikimedia.org/wikidatawiki/entities/20210215/ They can download the wikidata-20210215-all.json.bz2
> 2. Wikipedia dump. Example: https://dumps.wikimedia.org/enwiki/20210220/ they need to download enwiki-20210220-pages-articles-multistream.xml.bz2 and enwiki-20210220-pages-articles-multistream-index.txt.bz2
> 3. Pageview files, such as pagecounts-2020-08-views-ge-5-totals.bz2 from https://dumps.wikimedia.org/other/pagecounts-ez/merged/
> 
> You need to then supply the file names to the script. You can use the "--help" option to see it. If you run it on a single machine, Spark will also help with parallelizing it. If you run it on a single CPU it would be way too slow and may take days to preprocess. Let me know if you have other issues!

Hello, these links are no longer valid, can you please write the replacement links here. I looked at [1] but could not find the directory which has all the files needed.

1. https://dumps.wikimedia.org/wikidatawiki/entities/
